keps/sig-cli/4944-kustomize-wasm-plugin/README.md

```diff
@@ -0,0 +1,878 @@
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+This KEP proposes adding WebAssembly plugin support as a sandboxed execution
+environment for the Kubernetes Resource Model (KRM) framework in Kustomize.
+
+This enhancement enables platform-independent plugin execution in a secure,
+sandboxed environment, eliminating runtime dependencies while maintaining strict
+isolation.
+
+And promote kustomize KRM plugins to GA with the addition and stability of Wasm
+KRM Function.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+Kustomize currently supports various ways to transform Kubernetes resources,
+including built-in transformers and
+[container](https://kubectl.docs.kubernetes.io/guides/extending_kustomize/containerized_krm_functions/)/[exec](https://kubectl.docs.kubernetes.io/guides/extending_kustomize/exec_krm_functions/)
+plugins. However, these methods have limitations in terms of portability,
+security, and performance.
+
+By introducing Wasm plugin support, we can address these limitations and provide
+a more robust solution for custom resource transformations.
+
+#### Limitation current plugins
+
+The current Kustomize plugin ecosystem faces several operational challenges that
+limit its effectiveness in various environments. Container plugins, while
+providing good isolation, require a Docker runtime, which is often unavailable
+or restricted in CI/CD environments due to security policies or resource
+constraints. Exec plugins are tightly coupled to their underlying operating
+system and execution environment, requiring platform-specific compilation and
+distribution. This makes maintaining consistency across diverse environments
+challenging. Additionally, container plugins introduce architecture-specific
+dependencies, requiring builds for different CPU architectures (e.g., amd64,
+arm64), which complicates deployment across heterogeneous infrastructure. These
+limitations create scenarios where neither exec nor container plugins can be
+effectively utilized, particularly in organizations with diverse infrastructure
+requirements or strict security policies.
+
+### Goals
+
+1. Implement Wasm plugin support in the Kustomize KRM framework
+1. Provide a secure sandboxed environment for executing custom transformations
+1. Improve portability of custom transformers across different platforms
+1. Enhance performance of resource transformations through WebAssembly
```

Comment:
Nowhere in this document there's a clear information how this new mechanism will affect the kustomize version that is available in kubectl? IIRC the KRM functions are NOT available in kubectl, so I believe this one is also not going to be available, is that right assumption?

Reply:
Honesty, Exec KRM functions is disabled by kubectl, but the Container KRM functions is available for kustomize in kubectl now.
(We can use with `kubectl kustomize --enable-alpha-plugins ...`)
https://github.com/kubernetes-sigs/kustomize/issues/4556#issuecomment-1092095023

According to my understanding, the Exec KRM functions mechanism, which directly executes binaries, raises security concerns and was disabled in kubectl. With the Wasm Plugin, since the execution environment is sandboxed, I thought it seemed reasonable to make it available in the same way as current Container functions.